[inductor] Support integer pow

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88938

This allows the `pow_recursive` form to be used for any integer power
greater than 0, or for tensor cases to fallback to ATen.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
WIP: Support for integer pow

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Support for integer pow"




[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Support for integer pow"




[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Support for integer pow"




[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Support for integer pow"




[ghstack-poisoned]

peterbell10 committed:
Fix lint on "WIP: Support for integer pow"




[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Support for integer pow"




[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: Support for integer pow"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Support integer pow on CUDA"


This implements the "exponentiation by squaring" method used in `pow_recursive`
but with the conditionals evaluated on device at runtime. Since we don't have
looping control flow, this uses a fixed number of iterations depending on the
type of the exponent e.g. 8 for uint8, 63 for int64.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Support integer pow"


This implements the "exponentiation by squaring" method used in `pow_recursive`
but with the conditionals evaluated on device at runtime. Since we don't have
looping control flow, this uses a fixed number of iterations depending on the
type of the exponent e.g. 8 for uint8, 63 for int64.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Support integer pow"


This allows the `pow_recursive` form to be used for any integer power
greater than 0, or for tensor cases to fallback to ATen.

[ghstack-poisoned]

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -3530,29 +3530,25 @@ def pow_native(a, b):
+    is_integer_pow = is_integer_dtype(dtype)
+    # Optimize away small fixed powers, or for integers inline the simple cases
+    embed_exponent = isinstance(b, int) and (
+        -32 < b < 32 or (is_integer_pow and b >= 0)
```
What's the rationale behind `is_integer_pow and b >= 0`? Since `a` and `b` cannot both be scalars, I guess that `dtype` here is coming from `a`. If this is the case, is it alright that we would want to inline the computation of `a ** 12309481`? And if so, why don't we want to inline that of `a ** -12309481` as `1. / (a ** 12309481)`?

peterbell10 committed:
Rebase on "[inductor] Support integer pow"


This allows the `pow_recursive` form to be used for any integer power
greater than 0, or for tensor cases to fallback to ATen.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Support integer pow"


This allows the `pow_recursive` form to be used for any integer power
greater than 0, or for tensor cases to fallback to ATen.

[ghstack-poisoned]

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -3541,29 +3541,27 @@ def pow_native(a, b):
-    elif isinstance(b, int) and -32 < b < 32:
-        # Optimize away small fixed powers
+
+    # Type promotion ensures all tensor arguments have the same type
+    dtype = next(x.get_dtype() for x in (a, b) if isinstance(x, ir.TensorBox))
```
Nice!

peterbell10 commented:
@pytorchbot merge
